### PR TITLE
Uroboros

### PR DESCRIPTION
**Pull request by BritikovKI**
Implemented MacMillan paper about the interpolant chc approach

### CODE REVIEW

File: src/engine/Uroboros.cc [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+PTRef Uroboros::lastIterationInterpolant(MainSolver& solver) {
+    auto itpContext = solver.getInterpolationContext();
+    vec<PTRef> itps;
+    ipartitions_t mask = 1;
+    itpContext->getSingleInterpolant(itps, mask);
+    return itps[0];
+}
+
+sstat Uroboros::checkItp(PTRef & itp, PTRef & itpsOld, Logic & logic){
```
Comment (blishko) [inline src/engine/Uroboros.cc:80]: `PTRef` should be passed by value.

**Review by blishko — COMMENTED** [review]
Small comments from the first glance at the code.

Can you rebase on top of the current master and squash the first commits, so that the history does not contain the addition and deletion of the `build` and `.idea` folders?

**Review by blishko — COMMENTED** [review]
It seems to me the algorithm is OK, the code need some polishing, but then we can test it on larger benchmarks.

**Comment by blishko** [comment]
I think the stuff related to GitHub actions should be moved to separate PR.

Also, I think we have to rename the engine to `IMC`, that seems to be pretty standard when people refer to this McMillan's algorithm.

File: src/engine/Uroboros.cc (hunk 1 above)
Comment (BritikovKI) [reply src/engine/Uroboros.cc]: Resolved

File: src/engine/Uroboros.cc [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    std::size_t maxLoopUnrollings = std::numeric_limits<std::size_t>::max();
+    PTRef init = system.getInit();
+    PTRef query = system.getQuery();
+    PTRef transition = system.getTransition();
+
+    SMTConfig config;
+    const char * msg = "ok";
+    config.setOption(SMTConfig::o_produce_inter, SMTOption(true), msg);
+    config.setSimplifyInterpolant(4);
+    MainSolver solver(logic, config, "Uroboros");
```
[the remark this answers is not in the document]
Comment (BritikovKI) [reply src/engine/Uroboros.cc]: Resolved

File: src/engine/Uroboros.cc [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    Logic & initLogic = logic;
```
[the remark this answers is not in the document]
Comment (BritikovKI) [reply src/engine/Uroboros.cc]: Resolved

File: src/engine/Uroboros.cc [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            return GraphVerificationResult{VerificationResult::UNSAFE,  InvalidityWitness::fromTransitionSystem(graph, k)};
+        } if (res == l_False){
+            return GraphVerificationResult{VerificationResult::SAFE, ValidityWitness::fromTransitionSystem(logic, graph, system, init)};
+        }
+    }
+    return GraphVerificationResult(VerificationResult::UNKNOWN);
+}
+
+//procedure FiniteRun(M=(I,T,F), k>0)
+lbool Uroboros::finiteRun(PTRef & init, PTRef & transition, PTRef & query, int & k) {
```
[the remark this answers is not in the document]
Comment (BritikovKI) [reply src/engine/Uroboros.cc]: Resolved

**Comment by blishko** [comment]
Thanks!

**Pull request merged**